File: chaeyun/3_palindrome/palindromic-substr-cnt/palindromic-substr-dp.py
"""
dp[i][j] to indicate whether the substring from index i to index j is a palindrome or not

Approach : 
dp[i][j] = (s[i] == s[j]) and dp[i+1][j-1]
"""

class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        count = 0
        
        # base cases
        for i in range(n):
            dp[i][i] = True
            count += 1
        
        for i in range(n-1):
            dp[i][i+1] = (s[i] == s[i+1])
            count += dp[i][i+1]
        
        # recurrence relation
        for l in range(3, n+1):
            for i in range(n-l+1):
                j = i + l - 1
                dp[i][j] = (s[i] == s[j]) and dp[i+1][j-1]
                count += dp[i][j]
        
        return count

File: chaeyun/3_palindrome/palindromic-substr-cnt/palindromic-substr-dp.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int c=0,n=s.size();
        vector<vector<int>>dp(n,vector<int>(n,0));
        for(int diff=0;diff<n;diff++)
        {
            for(int i=0,j=i+diff; j<n; i++,j++)
            {
                if(diff==0)
                    dp[i][j]=1;
                 else if(diff==1)
                   {
                       dp[i][j]=(s[i]==s[j])?2:0;
                   }    
                else 
                {
                    if(s[i]==s[j] && dp[i+1][j-1])
                        dp[i][j]=dp[i+1][j-1]+2;

                }
                if(dp[i][j])
                    c++;
            }
        }
        return c;
    }
};